internal/commands/build/build.go

```diff
@@ -0,0 +1,240 @@
+
+	return cmd
+}
+
+func runBuild(dockerCli command.Cli, application string, opt buildOptions) (reference.Named, error) {
+	err := checkMinimalEngineVersion(dockerCli)
+	if err != nil {
+		return nil, err
+	}
+
+	var ref reference.Named
+	ref, err = packager.GetNamedTagged(opt.tag)
+	if err != nil {
+		return nil, err
+	}
+
+	app, err := packager.Extract(application)
+	if err != nil {
+		return nil, err
+	}
+	defer app.Cleanup()
+
+	bundle, err := packager.MakeBundleFromApp(dockerCli, app, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	buildopts, err := parseCompose(app, opt)
+	if err != nil {
+		return nil, err
+	}
+
+	buildopts["invocation-image"], err = createInvocationImageBuildOptions(dockerCli, app)
+	if err != nil {
+		return nil, err
+	}
+
+	debugBuildOpts(buildopts)
+
+	ctx, cancel := context.WithCancel(appcontext.Context())
+	defer cancel()
+	d, err := driver.GetDriver(ctx, "buildx_buildkit_default", nil, dockerCli.Client(), nil, "", nil)
+	if err != nil {
+		return nil, err
+	}
+	driverInfo := []build.DriverInfo{
+		{
+			Name:   "default",
+			Driver: d,
+		},
+	}
+
+	pw := progress.NewPrinter(ctx, os.Stderr, opt.progress)
+
+	// We rely on buildx "docker" builder integrated in docker engine, so don't nee a DockerAPI here
+	resp, err := build.Build(ctx, driverInfo, buildopts, nil, dockerCli.ConfigFile(), pw)
+	if err != nil {
+		return nil, err
+	}
+
+	fmt.Println("Successfully built service images")
+	updateBundle(bundle, resp)
+
+	if ref == nil {
+		if ref, err = computeDigest(bundle); err != nil {
+			return nil, err
+		}
+	}
+
+	if err = persistBundle(opt, bundle, ref); err != nil {
+		return nil, err
+	}
+	return ref, nil
+}
+
+func checkMinimalEngineVersion(dockerCli command.Cli) error {
+	info, err := dockerCli.Client().Info(appcontext.Context())
+	if err != nil {
+		return err
+	}
+	majorVersion, err := strconv.Atoi(info.ServerVersion[:strings.IndexRune(info.ServerVersion, '.')])
+	if err != nil {
+		return err
+	}
+	if majorVersion < 19 {
+		return errors.New("'build' require docker engine 19.03 or later")
+	}
+	return nil
+}
+
+func updateBundle(bundle *bundle.Bundle, resp map[string]*client.SolveResponse) {
+	debugSolveResponses(resp)
+	for service, r := range resp {
+		digest := r.ExporterResponse["containerimage.digest"]
+		if service == "invocation-image" {
+			bundle.InvocationImages[0].Digest = digest
+		} else {
+			image := bundle.Images[service]
+			image.Image = fmt.Sprintf("%s:%s-%s", bundle.Name, bundle.Version, service)
+			image.ImageType = cnab.ImageTypeDocker
+			image.Digest = digest
+			bundle.Images[service] = image
+		}
+	}
+	debugBundle(bundle)
+}
+
+func persistBundle(opt buildOptions, bndl *bundle.Bundle, ref reference.Named) error {
+	if opt.out != "" {
+		b, err := json.MarshalIndent(bndl, "", "  ")
+		if err != nil {
+			return err
+		}
+		if opt.out == "-" {
+			_, err = os.Stdout.Write(b)
+			if err != nil {
```

Comment:
```go
_, err = os.Stdout.Write(b)
return err
```

Reply:
I'm fine with early returns, just unclear to me if this is a recommended go usage

Reply:
https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow
It looks like it is in the golang repo itself 😄